File: internal/services/pubsub/validation.go
package pubsub

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"time"
)

var resourceIDPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._~+%-]{0,254}$`)

var attributeComparisonFilterPattern = regexp.MustCompile(`^attributes\.([A-Za-z0-9_.-]+)\s*(!=|=)\s*"([^"]*)"$`)

var attributePrefixFilterPattern = regexp.MustCompile(`^hasPrefix\(\s*attributes\.([A-Za-z0-9_.-]+)\s*,\s*"([^"]*)"\s*\)$`)

func validResourceID(id string) bool {
	return resourceIDPattern.MatchString(id)
}

func validProjectID(id string) bool {
	return resourceIDPattern.MatchString(id)
}

func validFullTopicName(name string) bool {
	parts := strings.Split(name, "/")
	return len(parts) == 4 && parts[0] == "projects" && parts[2] == "topics" && validProjectID(parts[1]) && validResourceID(parts[3])
}

func validFullSubscriptionName(name string) bool {
	parts := strings.Split(name, "/")
	return len(parts) == 4 && parts[0] == "projects" && parts[2] == "subscriptions" && validProjectID(parts[1]) && validResourceID(parts[3])
}

func validFullSnapshotName(name string) bool {
	parts := strings.Split(name, "/")
	return len(parts) == 4 && parts[0] == "projects" && parts[2] == "snapshots" && validProjectID(parts[1]) && validResourceID(parts[3])
}

func validFullSchemaName(name string) bool {
	parts := strings.Split(name, "/")
	return len(parts) == 4 && parts[0] == "projects" && parts[2] == "schemas" && validProjectID(parts[1]) && validResourceID(parts[3])
}

func validSchemaType(schemaType string) bool {
	switch schemaType {
	case "", "TYPE_UNSPECIFIED", "PROTOCOL_BUFFER", "AVRO":
		return true
	default:
		return false
	}
}

func validSchemaEncoding(encoding string) bool {
	switch encoding {
	case "", "ENCODING_UNSPECIFIED", "JSON", "BINARY":
		return true
	default:
		return false
	}
}

func validSchemaMessageData(message []byte, encoding string) bool {
	if len(message) == 0 {
		return true
	}
	switch encoding {
	case "JSON":
		return json.Valid(message)
	default:
		return true
	}
}

func validateSchemaDefinition(schemaType string, definition string) error {
	if strings.TrimSpace(definition) == "" {
		return nil
	}
	switch schemaType {
	case "AVRO":
		var decoded any
		if err := json.Unmarshal([]byte(definition), &decoded); err != nil {
			return errors.New("avro schema definition must be valid json")
		}
		if _, ok := decoded.(map[string]any); !ok {
			return errors.New("avro schema definition must be a json object")
		}
	}
	return nil
}

func emptySchemaResource(schema schemaResource) bool {
	return schema.Name == "" &&
		schema.Type == "" &&
		schema.Definition == "" &&
		schema.RevisionID == "" &&
		schema.RevisionCreateTime == "" &&
		len(schema.Revisions) == 0
}

func validateDeadLetterPolicy(policy map[string]any) error {
	if len(policy) == 0 {
		return nil
	}
	rawTopic, ok := policy["deadLetterTopic"]
	if !ok {
		return fmt.Errorf("deadLetterPolicy.deadLetterTopic is required")
	}
	topic, ok := rawTopic.(string)
	if !ok || !validFullTopicName(topic) {
		return fmt.Errorf("invalid deadLetterPolicy.deadLetterTopic")
	}
	maxAttempts, ok := deadLetterMaxDeliveryAttempts(policy)
	if !ok {
		return fmt.Errorf("deadLetterPolicy.maxDeliveryAttempts is required")
	}
	if maxAttempts < 5 || maxAttempts > 100 {
		return fmt.Errorf("deadLetterPolicy.maxDeliveryAttempts must be between 5 and 100")
	}
	return nil
}

func (s *Server) deadLetterTopicExistsLocked(policy map[string]any) bool {
	if len(policy) == 0 {
		return true
	}
	_, found := s.topics[deadLetterTopic(policy)]
	return found
}

func validateTopicMetadata(topic topicResource) error {
	if strings.TrimSpace(topic.MessageRetentionDuration) != "" {
		if _, err := parseGoogleDuration(topic.MessageRetentionDuration); err != nil {
			return fmt.Errorf("messageRetentionDuration must be a non-negative duration")
		}
	}
	if len(topic.SchemaSettings) > 0 {
		rawSchema, ok := topic.SchemaSettings["schema"]
		if !ok {
			return fmt.Errorf("schemaSettings.schema is required")
		}
		schema, ok := rawSchema.(string)
		if !ok || !validFullSchemaName(schema) {
			return fmt.Errorf("invalid schemaSettings.schema")
		}
		if rawEncoding, ok := topic.SchemaSettings["encoding"]; ok {
			encoding, ok := rawEncoding.(string)
			if !ok || !validSchemaEncoding(encoding) {
				return fmt.Errorf("invalid schemaSettings.encoding")
			}
		}
	}
	return nil
}

func validateSubscriptionMetadata(subscription subscriptionResource) error {
	if strings.TrimSpace(subscription.MessageRetentionDuration) != "" {
		if _, err := parseGoogleDuration(subscription.MessageRetentionDuration); err != nil {
			return fmt.Errorf("messageRetentionDuration must be a non-negative duration")
		}
	}
	if len(subscription.ExpirationPolicy) > 0 {
		rawTTL, ok := subscription.ExpirationPolicy["ttl"]
		if !ok {
			return fmt.Errorf("expirationPolicy.ttl is required")
		}
		ttl, ok := rawTTL.(string)
		if !ok || strings.TrimSpace(ttl) == "" {
			return fmt.Errorf("expirationPolicy.ttl must be a duration string")
		}
		if _, err := parseGoogleDuration(ttl); err != nil {
			return fmt.Errorf("expirationPolicy.ttl must be a non-negative duration")
		}
	}
	return nil
}

func parseGoogleDuration(value string) (time.Duration, error) {
	value = strings.TrimSpace(value)
	if value == "" {
		return 0, fmt.Errorf("empty duration")
	}
	duration, err := time.ParseDuration(value)
	if err != nil || duration < 0 {
		return 0, fmt.Errorf("invalid duration")
	}
	return duration, nil
}

func validateSubscriptionFilter(filter string) error {
	filter = strings.TrimSpace(filter)
	if filter == "" {
		return nil
	}
	if !attributeComparisonFilterPattern.MatchString(filter) && !attributePrefixFilterPattern.MatchString(filter) {
		return fmt.Errorf("unsupported subscription filter")
	}
	return nil
}

func validateRetryPolicy(policy map[string]any) error {
	if len(policy) == 0 {
		return nil
	}
	minimum, hasMinimum, err := retryPolicyDuration(policy, "minimumBackoff")
	if err != nil {
		return err
	}
	maximum, hasMaximum, err := retryPolicyDuration(policy, "maximumBackoff")
	if err != nil {
		return err
	}
	if hasMinimum && hasMaximum && minimum > maximum {
		return fmt.Errorf("retryPolicy.minimumBackoff must be less than or equal to retryPolicy.maximumBackoff")
	}
	return nil
}

func validatePushConfig(config map[string]any) error {
	if len(config) == 0 {
		return nil
	}
	rawEndpoint, ok := config["pushEndpoint"]
	if !ok {
		return nil
	}
	endpoint, ok := rawEndpoint.(string)
	if !ok || strings.TrimSpace(endpoint) == "" {
		return fmt.Errorf("pushConfig.pushEndpoint must be an http or https URL")
	}
	parsed, err := url.Parse(endpoint)
	if err != nil || parsed.Host == "" {
		return fmt.Errorf("pushConfig.pushEndpoint must be an http or https URL")
	}
	if parsed.Scheme != "http" && parsed.Scheme != "https" {
		return fmt.Errorf("pushConfig.pushEndpoint must be an http or https URL")
	}
	if parsed.User != nil {
		return fmt.Errorf("pushConfig.pushEndpoint must not include user info")
	}
	return nil
}

func validatePublishMessage(data string, attributes map[string]string) error {
	if data == "" && len(attributes) == 0 {
		return fmt.Errorf("message data or attributes are required")
	}
	if data != "" {
		if _, err := decodeBase64Bytes(data); err != nil {
			return fmt.Errorf("message data must be base64 encoded")
		}
	}
	for key := range attributes {
		if strings.TrimSpace(key) == "" {
			return fmt.Errorf("message attributes must not contain empty keys")
		}
	}
	return nil
}

func validateMessageAgainstTopicSchemaSettings(data string, schemaSettings map[string]any) error {
	if len(schemaSettings) == 0 {
		return nil
	}
	encoding, _ := schemaSettings["encoding"].(string)
	if encoding == "" {
		return nil
	}
	message, err := decodeBase64Bytes(data)
	if err != nil {
		return fmt.Errorf("message data must be base64 encoded")
	}
	if !validSchemaMessageData(message, encoding) {
		return fmt.Errorf("message is invalid for topic schema encoding")
	}
	return nil
}

func decodeBase64Bytes(value string) ([]byte, error) {
	for _, encoding := range []*base64.Encoding{
		base64.StdEncoding,
		base64.RawStdEncoding,
	} {
		decoded, err := encoding.DecodeString(value)
		if err == nil {
			return decoded, nil
		}
	}
	return nil, fmt.Errorf("invalid base64")
}

func retryPolicyDuration(policy map[string]any, field string) (time.Duration, bool, error) {
	raw, ok := policy[field]
	if !ok {
		return 0, false, nil
	}
	value, ok := raw.(string)
	if !ok || strings.TrimSpace(value) == "" {
		return 0, false, fmt.Errorf("retryPolicy.%s must be a duration string", field)
	}
	duration, err := time.ParseDuration(value)
	if err != nil || duration < 0 {
		return 0, false, fmt.Errorf("retryPolicy.%s must be a non-negative duration", field)
	}
	return duration, true, nil
}

func subscriptionMatchesMessage(subscription subscriptionResource, message pubsubMessage) bool {
	filter := strings.TrimSpace(subscription.Filter)
	if filter == "" {
		return true
	}
	if match := attributeComparisonFilterPattern.FindStringSubmatch(filter); len(match) == 4 {
		value := message.Attributes[match[1]]
		if match[2] == "!=" {
			return value != match[3]
		}
		return value == match[3]
	}
	match := attributePrefixFilterPattern.FindStringSubmatch(filter)
	if len(match) != 3 {
		return false
	}
	return strings.HasPrefix(message.Attributes[match[1]], match[2])
}

func deadLetterMaxDeliveryAttempts(policy map[string]any) (int, bool) {
	raw, ok := policy["maxDeliveryAttempts"]
	if !ok {
		return 0, false
	}
	switch value := raw.(type) {
	case int:
		return value, true
	case int64:
		return int(value), true
	case float64:
		if value != float64(int(value)) {
			return 0, false
		}
		return int(value), true
	case json.Number:
		parsed, err := value.Int64()
		if err != nil {
			return 0, false
		}
		return int(parsed), true
	default:
		return 0, false
	}
}

func deadLetterTopic(policy map[string]any) string {
	if len(policy) == 0 {
		return ""
	}
	topic, _ := policy["deadLetterTopic"].(string)
	return topic
}

func resourceProject(name string) string {
	parts := strings.Split(name, "/")
	if len(parts) >= 2 && parts[0] == "projects" {
		return parts[1]
	}
	return ""
}

File: services/pubsub/src/validation.rs
//! Subscription metadata / policy validation.
//!
//! Mirrors the subscription-side validators in
//! `internal/services/pubsub/validation.rs` (filter, retry policy, push config,
//! dead-letter policy, expiration policy, message-retention duration). Error
//! wording is verbatim. Filter matching uses the same two patterns legacy compiles:
//! an attribute comparison and a `hasPrefix(...)` form.

use serde_json::Value;

use crate::duration::valid_google_duration;
use crate::errors::ApiError;
use crate::paths::valid_full_topic_name;

/// True for `""`/`TYPE_UNSPECIFIED`/`PROTOCOL_BUFFER`/`AVRO`, mirroring
/// `validSchemaType`.
pub fn valid_schema_type(schema_type: &str) -> bool {
    matches!(
        schema_type,
        "" | "TYPE_UNSPECIFIED" | "PROTOCOL_BUFFER" | "AVRO"
    )
}

/// True for `""`/`ENCODING_UNSPECIFIED`/`JSON`/`BINARY`, mirroring
/// `validSchemaEncoding`.
pub fn valid_schema_encoding(encoding: &str) -> bool {
    matches!(encoding, "" | "ENCODING_UNSPECIFIED" | "JSON" | "BINARY")
}

/// Validates a schema definition, mirroring `validateSchemaDefinition` (AVRO
/// must be a JSON object; others are unchecked). An empty/whitespace definition
/// is always allowed.
pub fn validate_schema_definition(schema_type: &str, definition: &str) -> Result<(), ApiError> {
    if definition.trim().is_empty() {
        return Ok(());
    }
    if schema_type == "AVRO" {
        match serde_json::from_str::<Value>(definition) {
            Ok(Value::Object(_)) => {}
            Ok(_) => {
                return Err(ApiError::invalid_argument(
                    "avro schema definition must be a json object",
                ))
            }
            Err(_) => {
                return Err(ApiError::invalid_argument(
                    "avro schema definition must be valid json",
                ))
            }
        }
    }
    Ok(())
}

/// Validates message bytes for a schema encoding, mirroring
/// `validSchemaMessageData` (only `JSON` is checked: it must be valid JSON).
pub fn valid_schema_message_data(message: &[u8], encoding: &str) -> bool {
    if message.is_empty() {
        return true;
    }
    if encoding == "JSON" {
        return serde_json::from_slice::<Value>(message).is_ok();
    }
    true
}

/// Decodes base64 trying StdEncoding then RawStdEncoding, mirroring
/// `decodeBase64Bytes`.
pub fn decode_base64_bytes(value: &str) -> Option<Vec<u8>> {
    base64_decode(value, true).or_else(|| base64_decode(value, false))
}

/// Minimal base64 decoder. `padded` selects StdEncoding (requires `=` padding to
/// a multiple of 4) vs RawStdEncoding (no padding).
fn base64_decode(input: &str, padded: bool) -> Option<Vec<u8>> {
    fn val(c: u8) -> Option<u32> {
        match c {
            b'A'..=b'Z' => Some((c - b'A') as u32),
            b'a'..=b'z' => Some((c - b'a' + 26) as u32),
            b'0'..=b'9' => Some((c - b'0' + 52) as u32),
            b'+' => Some(62),
            b'/' => Some(63),
            _ => None,
        }
    }
    let bytes = input.as_bytes();
    if padded {
        if !bytes.len().is_multiple_of(4) {
            return None;
        }
        let mut out = Vec::with_capacity(bytes.len() / 4 * 3);
        for chunk in bytes.chunks(4) {
            let pads = chunk.iter().rev().take_while(|&&b| b == b'=').count();
            if pads > 2 {
                return None;
            }
            let mut acc = 0u32;
            for (j, &c) in chunk.iter().enumerate() {
                let v = if c == b'=' {
                    if j < 4 - pads {
                        return None;
                    }
                    0
                } else {
                    val(c)?
                };
                acc = (acc << 6) | v;
            }
            out.push((acc >> 16) as u8);
            if pads < 2 {
                out.push((acc >> 8) as u8);
            }
            if pads < 1 {
                out.push(acc as u8);
            }
        }
        Some(out)
    } else {
        // RawStdEncoding: no padding allowed.
        if bytes.contains(&b'=') {
            return None;
        }
        let mut out = Vec::with_capacity(bytes.len() * 3 / 4);
        for chunk in bytes.chunks(4) {
            if chunk.len() == 1 {
                return None;
            }
            let mut acc = 0u32;
            for &c in chunk {
                acc = (acc << 6) | val(c)?;
            }
            acc <<= 6 * (4 - chunk.len());
            out.push((acc >> 16) as u8);
            if chunk.len() > 2 {
                out.push((acc >> 8) as u8);
            }
            if chunk.len() > 3 {
                out.push(acc as u8);
            }
        }
        Some(out)
    }
}

/// Validates a subscription filter, mirroring `validateSubscriptionFilter`.
pub fn validate_subscription_filter(filter: &str) -> Result<(), ApiError> {
    let filter = filter.trim();
    if filter.is_empty() {
        return Ok(());
    }
    if parse_comparison_filter(filter).is_none() && parse_prefix_filter(filter).is_none() {
        return Err(ApiError::invalid_argument(
            "unsupported subscription filter",
        ));
    }
    Ok(())
}

/// Parses `attributes.<key> [!=|=] "<value>"`, returning `(key, op, value)`.
/// Mirrors `attributeComparisonFilterPattern`:
/// `^attributes\.([A-Za-z0-9_.-]+)\s*(!=|=)\s*"([^"]*)"$`.
pub fn parse_comparison_filter(filter: &str) -> Option<(String, String, String)> {
    let rest = filter.strip_prefix("attributes.")?;
    // key = [A-Za-z0-9_.-]+
    let key_end = rest
        .find(|c: char| !(c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-')))
        .unwrap_or(rest.len());
    if key_end == 0 {
        return None;
    }
    let key = &rest[..key_end];
    let after_key = rest[key_end..].trim_start();
    let (op, after_op) = if let Some(r) = after_key.strip_prefix("!=") {
        ("!=", r)
    } else if let Some(r) = after_key.strip_prefix('=') {
        ("=", r)
    } else {
        return None;
    };
    let after_op = after_op.trim_start();
    let value = after_op.strip_prefix('"')?;
    let value = value.strip_suffix('"')?;
    // `[^"]*` — the value must contain no embedded quote.
    if value.contains('"') {
        return None;
    }
    Some((key.to_string(), op.to_string(), value.to_string()))
}

/// Parses `hasPrefix( attributes.<key> , "<prefix>" )`, returning `(key, prefix)`.
/// Mirrors `attributePrefixFilterPattern`.
pub fn parse_prefix_filter(filter: &str) -> Option<(String, String)> {
    let inner = filter.strip_prefix("hasPrefix(")?.strip_suffix(')')?;
    let inner = inner.trim();
    let inner = inner.strip_prefix("attributes.")?;
    let (key_part, value_part) = inner.split_once(',')?;
    let key = key_part.trim();
    if key.is_empty()
        || !key
            .chars()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, '_' | '.' | '-'))
    {
        return None;
    }
    let value_part = value_part.trim();
    let value = value_part.strip_prefix('"')?.strip_suffix('"')?;
    if value.contains('"') {
        return None;
    }
    Some((key.to_string(), value.to_string()))
}

/// Validates `messageRetentionDuration` + `expirationPolicy`, mirroring
/// `validateSubscriptionMetadata`.
pub fn validate_subscription_metadata(
    message_retention_duration: &str,
    expiration_policy: Option<&Value>,
) -> Result<(), ApiError> {
    if !message_retention_duration.trim().is_empty()
        && !valid_google_duration(message_retention_duration)
    {
        return Err(ApiError::invalid_argument(
            "messageRetentionDuration must be a non-negative duration",
        ));
    }
    if let Some(policy) = expiration_policy.and_then(Value::as_object) {
        if policy.is_empty() {
            return Ok(());
        }
        let ttl = policy
            .get("ttl")
            .ok_or_else(|| ApiError::invalid_argument("expirationPolicy.ttl is required"))?;
        let ttl = match ttl.as_str() {
            Some(s) if !s.trim().is_empty() => s,
            _ => {
                return Err(ApiError::invalid_argument(
                    "expirationPolicy.ttl must be a duration string",
                ))
            }
        };
        if !valid_google_duration(ttl) {
            return Err(ApiError::invalid_argument(
                "expirationPolicy.ttl must be a non-negative duration",
            ));
        }
    }
    Ok(())
}

/// Validates a dead-letter policy, mirroring `validateDeadLetterPolicy`.
pub fn validate_dead_letter_policy(policy: Option<&Value>) -> Result<(), ApiError> {
    let Some(policy) = policy.and_then(Value::as_object) else {
        return Ok(());
    };
    if policy.is_empty() {
        return Ok(());
    }
    let topic = policy.get("deadLetterTopic").ok_or_else(|| {
        ApiError::invalid_argument("deadLetterPolicy.deadLetterTopic is required")
    })?;
    match topic.as_str() {
        Some(t) if valid_full_topic_name(t) => {}
        _ => {
            return Err(ApiError::invalid_argument(
                "invalid deadLetterPolicy.deadLetterTopic",
            ))
        }
    }
    let max_attempts = dead_letter_max_delivery_attempts(policy).ok_or_else(|| {
        ApiError::invalid_argument("deadLetterPolicy.maxDeliveryAttempts is required")
    })?;
    if !(5..=100).contains(&max_attempts) {
        return Err(ApiError::invalid_argument(
            "deadLetterPolicy.maxDeliveryAttempts must be between 5 and 100",
        ));
    }
    Ok(())
}

/// The dead-letter topic name from a policy (empty when absent), mirroring
/// `deadLetterTopic`.
pub fn dead_letter_topic(policy: Option<&Value>) -> String {
    policy
        .and_then(Value::as_object)
        .and_then(|o| o.get("deadLetterTopic"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string()
}

fn dead_letter_max_delivery_attempts(policy: &serde_json::Map<String, Value>) -> Option<i64> {
    let raw = policy.get("maxDeliveryAttempts")?;
    match raw {
        Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                Some(i)
            } else {
                // float that is integral
                let f = n.as_f64()?;
                if f == f.trunc() {
                    Some(f as i64)
                } else {
                    None
                }
            }
        }
        _ => None,
    }
}

/// Validates a retry policy, mirroring `validateRetryPolicy`.
pub fn validate_retry_policy(policy: Option<&Value>) -> Result<(), ApiError> {
    let Some(policy) = policy.and_then(Value::as_object) else {
        return Ok(());
    };
    if policy.is_empty() {
        return Ok(());
    }
    let min = retry_policy_duration(policy, "minimumBackoff")?;
    let max = retry_policy_duration(policy, "maximumBackoff")?;
    if let (Some(min), Some(max)) = (min, max) {
        if min > max {
            return Err(ApiError::invalid_argument(
                "retryPolicy.minimumBackoff must be less than or equal to retryPolicy.maximumBackoff",
            ));
        }
    }
    Ok(())
}

fn retry_policy_duration(
    policy: &serde_json::Map<String, Value>,
    field: &str,
) -> Result<Option<i128>, ApiError> {
    let Some(raw) = policy.get(field) else {
        return Ok(None);
    };
    let value = match raw.as_str() {
        Some(s) if !s.trim().is_empty() => s,
        _ => {
            return Err(ApiError::invalid_argument(format!(
                "retryPolicy.{field} must be a duration string"
            )))
        }
    };
    match crate::duration::parse_go_duration(value) {
        Some(n) if n >= 0 => Ok(Some(n)),
        _ => Err(ApiError::invalid_argument(format!(
            "retryPolicy.{field} must be a non-negative duration"
        ))),
    }
}

/// Validates a push config, mirroring `validatePushConfig`.
pub fn validate_push_config(config: Option<&Value>) -> Result<(), ApiError> {
    let Some(config) = config.and_then(Value::as_object) else {
        return Ok(());
    };
    if config.is_empty() {
        return Ok(());
    }
    let Some(raw_endpoint) = config.get("pushEndpoint") else {
        return Ok(());
    };
    let endpoint = match raw_endpoint.as_str() {
        Some(s) if !s.trim().is_empty() => s,
        _ => {
            return Err(ApiError::invalid_argument(
                "pushConfig.pushEndpoint must be an http or https URL",
            ))
        }
    };
    validate_push_endpoint(endpoint)
}

/// Minimal URL check matching legacy `url.Parse` usage: scheme http/https, a
/// non-empty host, and no user info.
fn validate_push_endpoint(endpoint: &str) -> Result<(), ApiError> {
    let bad = || ApiError::invalid_argument("pushConfig.pushEndpoint must be an http or https URL");
    let user_info_err =
        || ApiError::invalid_argument("pushConfig.pushEndpoint must not include user info");
    let (scheme, rest) = endpoint.split_once("://").ok_or_else(bad)?;
    if scheme != "http" && scheme != "https" {
        return Err(bad());
    }
    // Authority is up to the first '/', '?' or '#'.
    let authority_end = rest.find(['/', '?', '#']).unwrap_or(rest.len());
    let authority = &rest[..authority_end];
    let host = match authority.rsplit_once('@') {
        Some((user, _host)) if !user.is_empty() => return Err(user_info_err()),
        Some((_, host)) => host,
        None => authority,
    };
    if host.is_empty() {
        return Err(bad());
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn filter_patterns() {
        assert!(validate_subscription_filter("attributes.region = \"us\"").is_ok());
        assert!(validate_subscription_filter("attributes.x != \"y\"").is_ok());
        assert!(validate_subscription_filter("hasPrefix(attributes.k, \"pre\")").is_ok());
        assert!(validate_subscription_filter("bogus").is_err());
        assert_eq!(
            parse_comparison_filter("attributes.region = \"us\""),
            Some(("region".to_string(), "=".to_string(), "us".to_string()))
        );
    }

    #[test]
    fn dead_letter_policy_bounds() {
        assert!(validate_dead_letter_policy(Some(&json!({
            "deadLetterTopic": "projects/p/topics/dlq", "maxDeliveryAttempts": 5
        })))
        .is_ok());
        assert_eq!(
            validate_dead_letter_policy(Some(&json!({
                "deadLetterTopic": "projects/p/topics/dlq", "maxDeliveryAttempts": 2
            })))
            .unwrap_err()
            .message,
            "deadLetterPolicy.maxDeliveryAttempts must be between 5 and 100"
        );
    }

    #[test]
    fn retry_policy_order() {
        assert!(validate_retry_policy(Some(
            &json!({"minimumBackoff": "10s", "maximumBackoff": "600s"})
        ))
        .is_ok());
        assert!(validate_retry_policy(Some(
            &json!({"minimumBackoff": "600s", "maximumBackoff": "10s"})
        ))
        .is_err());
    }

    #[test]
    fn push_endpoint_checks() {
        assert!(
            validate_push_config(Some(&json!({"pushEndpoint": "https://example.com/push"})))
                .is_ok()
        );
        assert_eq!(
            validate_push_config(Some(&json!({"pushEndpoint": "ftp://bad"})))
                .unwrap_err()
                .message,
            "pushConfig.pushEndpoint must be an http or https URL"
        );
        assert_eq!(
            validate_push_config(Some(&json!({"pushEndpoint": "https://user@example.com/"})))
                .unwrap_err()
                .message,
            "pushConfig.pushEndpoint must not include user info"
        );
    }
}
